File: FlexServ Backend/DotNet/AdminService/Controllers/UserController.cs
using AdminService.Dtos;
using AdminService.Services;
using Microsoft.AspNetCore.Mvc;
using System;
using System.Collections.Generic;
using System.Security.Claims;

namespace AdminService.Controllers
{
    [ApiController]
    [Route("api/users")]
    [Route("api/user")]
    public class UserController : ControllerBase
    {
        private readonly IUserService _userService;

        public UserController(IUserService userService)
        {
            _userService = userService;
        }

        private long GetCurrentUserId()
        {
            var userIdClaim = User.FindFirstValue(ClaimTypes.NameIdentifier) ?? User.FindFirstValue("sub");
            if (long.TryParse(userIdClaim, out var userId))
            {
                return userId;
            }
            // Fallback for testing/unauthenticated calls
            return 1;
        }

        [HttpGet("profile")]
        public IActionResult GetProfile()
        {
            try
            {
                var userId = GetCurrentUserId();
                var user = _userService.GetUserProfile(userId);
                return Ok(new ApiResponse<UserResponse>(true, "Profile fetched successfully", user));
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new ApiResponse<UserResponse>(false, ex.Message, null));
            }
        }

        [HttpPut("profile")]
        public IActionResult UpdateProfile([FromBody] UpdateUserRequest request)
        {
            try
            {
                var userId = GetCurrentUserId();
                var updatedUser = _userService.UpdateUserProfile(userId, request);
                return Ok(new ApiResponse<UserResponse>(true, "Profile updated successfully", updatedUser));
            }
            catch (Exception ex) when (ex is KeyNotFoundException || ex is InvalidOperationException)
            {
                return BadRequest(new ApiResponse<UserResponse>(false, ex.Message, null));
            }
        }

        [HttpDelete("profile")]
        public IActionResult DeleteAccount()
        {
            try
            {
                var userId = GetCurrentUserId();
                _userService.DeleteUserAccount(userId);
                return Ok(new ApiResponse<string>(true, "Account deleted successfully", null));
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new ApiResponse<string>(false, ex.Message, null));
            }
        }

        [HttpPost("addresses")]
        public IActionResult AddAddress([FromBody] AddressRequest request)
        {
            try
            {
                var userId = GetCurrentUserId();
                var address = _userService.AddAddress(userId, request);
                return Created(string.Empty, new ApiResponse<AddressResponse>(true, "Address added successfully", address));
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new ApiResponse<AddressResponse>(false, ex.Message, null));
            }
        }

        [HttpGet("addresses")]
        public IActionResult GetAddresses()
        {
            try
            {
                var userId = GetCurrentUserId();
                var addresses = _userService.GetUserAddresses(userId);
                return Ok(new ApiResponse<List<AddressResponse>>(true, "Addresses fetched successfully", addresses));
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new ApiResponse<List<AddressResponse>>(false, ex.Message, null));
            }
        }

        [HttpPut("addresses/{addressId:long}")]
        public IActionResult UpdateAddress(long addressId, [FromBody] AddressRequest request)
        {
            try
            {
                var userId = GetCurrentUserId();
                var address = _userService.UpdateAddress(userId, addressId, request);
                return Ok(new ApiResponse<AddressResponse>(true, "Address updated successfully", address));
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new ApiResponse<AddressResponse>(false, ex.Message, null));
            }
        }

        [HttpDelete("addresses/{addressId:long}")]
        public IActionResult DeleteAddress(long addressId)
        {
            try
            {
                var userId = GetCurrentUserId();
                _userService.DeleteAddress(userId, addressId);
                return Ok(new ApiResponse<string>(true, "Address deleted successfully", null));
            }
            catch (KeyNotFoundException ex)
            {
                return NotFound(new ApiResponse<string>(false, ex.Message, null));
            }
        }
    }
}

File: FlexServ Backend/SpringBoot/BusinessService/src/main/java/com/flexserv/controller/UserController.java
package com.flexserv.controller;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.flexserv.dto.request.AddressRequest;
import com.flexserv.dto.request.UpdateUserRequest;
import com.flexserv.dto.response.AddressResponse;
import com.flexserv.dto.response.UserResponse;
import com.flexserv.payload.ApiResponse;
import com.flexserv.service.UserService;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;

@RestController
@RequestMapping("/api/users")
@RequiredArgsConstructor
public class UserController {

    private final UserService userService;

    // --- User Profile Endpoints ---

    @GetMapping("/profile")
    public ResponseEntity<ApiResponse<UserResponse>> getProfile(
            @RequestHeader(value = "X-User-Id", required = false) Long userId) {
        
        UserResponse user = userService.getUserProfile(userId);
        return ResponseEntity.ok(new ApiResponse<>(true, "Profile fetched successfully", user));
    }

    @PutMapping("/profile")
    public ResponseEntity<ApiResponse<UserResponse>> updateProfile(
            @RequestHeader(value = "X-User-Id", required = false) Long userId,
            @Valid @RequestBody UpdateUserRequest request) {

        UserResponse updatedUser = userService.updateUserProfile(userId, request);
        return ResponseEntity.ok(new ApiResponse<>(true, "Profile updated successfully", updatedUser));
    }

    @DeleteMapping("/profile")
    public ResponseEntity<ApiResponse<String>> deleteAccount(
            @RequestHeader(value = "X-User-Id", required = false) Long userId) {

        userService.deleteUserAccount(userId);
        return ResponseEntity.ok(new ApiResponse<>(true, "Account deleted successfully", null));
    }

    // --- Address Endpoints ---

    @PostMapping("/addresses")
    public ResponseEntity<ApiResponse<AddressResponse>> addAddress(
            @RequestHeader(value = "X-User-Id", required = false) Long userId,
            @Valid @RequestBody AddressRequest request) {

        AddressResponse address = userService.addAddress(userId, request);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(new ApiResponse<>(true, "Address added successfully", address));
    }

    @GetMapping("/addresses")
    public ResponseEntity<ApiResponse<List<AddressResponse>>> getAddresses(
            @RequestHeader(value = "X-User-Id", required = false) Long userId) {

        List<AddressResponse> addresses = userService.getUserAddresses(userId);
        return ResponseEntity.ok(new ApiResponse<>(true, "Addresses fetched successfully", addresses));
    }

    @PutMapping("/addresses/{addressId}")
    public ResponseEntity<ApiResponse<AddressResponse>> updateAddress(
            @RequestHeader(value = "X-User-Id", required = false) Long userId,
            @PathVariable Long addressId,
            @Valid @RequestBody AddressRequest request) {

        AddressResponse address = userService.updateAddress(userId, addressId, request);
        return ResponseEntity.ok(new ApiResponse<>(true, "Address updated successfully", address));
    }

    @DeleteMapping("/addresses/{addressId}")
    public ResponseEntity<ApiResponse<String>> deleteAddress(
            @RequestHeader(value = "X-User-Id", required = false) Long userId,
            @PathVariable Long addressId) {

        userService.deleteAddress(userId, addressId);
        return ResponseEntity.ok(new ApiResponse<>(true, "Address deleted successfully", null));
    }
}